synapse/handlers/pagination.py

```diff
@@ -132,6 +136,82 @@ def __init__(self, hs: "HomeServer"):
                     job.longest_max_lifetime,
                 )
 
+        if self._is_master:
+            self.clock.looping_call(
+                run_as_background_process,
+                PaginationHandler.PURGE_ROOMS_PERIOD,
+                "purge_rooms",
+                self.purge_rooms,
+            )
+
+    async def purge_rooms(self) -> None:
+        """This takes care of restoring unfinished purge/shutdown rooms from the DB.
+        It also takes care to launch scheduled ones, like rooms that has been fully
+        forgotten.
+
+        It should be run regularly.
+        """
+        rooms_to_purge = await self.store.get_rooms_to_purge()
```

Comment:
Are we worried about there being more rooms to purge than fit in memory ever? Should we have a limit? Perhaps the logic needs to incorporate `status = 'wait_purge' && now_ts > timestamp`

Otherwise, it's a pre-optimization

Reply:
I don't think we will hit any trouble here, since the delete jobs are run every hour, and subsequently cleaned.

I am planning to add a task to add all already forgotten rooms to this table on a later PR.
I'll take care of checking how many rooms we are talking on matrix.org here, but while sizeable it's probably more in the 10 of thousands than 10 of millions (I hope ^^).